### PR TITLE
Js modules

### PR DESCRIPTION
**Pull request by caroleolivier**

### CODE REVIEW

File: _posts/2017-11-21-javascript-modules.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
+---
+layout: post
+title: JavaScript Modules
+date: '2017-11-21'
+---
+
+Today, I looked a bit closer at JavaScript Modules. I have been using modules since I started using JavaScript but I must admit I have never looked into the subject deeply. Using [Babel](https://babeljs.io/) and [webpack](https://webpack.js.org/) things have been working fine and I just assumed webpack built a sort of dependency tree and Babel helped it understanding the latest ES module syntax (import/export).
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: Understand~ing~

File: _posts/2017-11-21-javascript-modules.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+<br/>
+This post is about what I learnt and understood about modules in the context of my setup (babel + webpack). To be honest this is from me to me, if you want to learn about modules and you happen to read this post, be critical.
+
+#### JavaScript Modules
+
+First of all, what is a module? The idea behind module is to isolate a piece of code so that it is easier to maintain, remove, modify, use. Modules are available natively in many languages but may be called differently: C# namespace, Java package, python modules, etc...
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: Maybe rephrase the last sentence to something like: "The concept of modules exist in many languages but might be called different things:..."  - it's a tricky sentence to make unambiguous because of the overloading of the word 'called' in programming :)

File: _posts/2017-11-21-javascript-modules.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+In the JavaScript language, modules have been available natively only since 2015 (wow!) with the ECMAScript 6 [standard](http://www.ecma-international.org/ecma-262/6.0/ECMA-262.pdf). Not being able to break down code into modules is madness (in my opinion) and the community has come up with other ways to add modules to the language. (I am not really sure why modules were not added before, I don't know what the process is to create a new ECMAScript standard).
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: Modules have _only_ been available natively ~only~ since

File: _posts/2017-11-21-javascript-modules.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+<br/>
+The most famous module standards (pre ES6) are [AMD](https://github.com/amdjs/amdjs-api/blob/master/AMD.md)
+(a browser targeted standard whose most famous implementation is [requireJS](http://requirejs.org/)) and [CommonJS Modules](http://wiki.commonjs.org/wiki/Modules/1.1) (a standard defined by the group [CommonJS](https://en.wikipedia.org/wiki/CommonJS)).
+
+So depending on the context, people may refer to ES modules (ESM), CommonJS (CJS) modules or AMD (I don't think there is even a consistency in acronym 🤔).
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: ~a~ consistency in acronym_s_

File: _posts/2017-11-21-javascript-modules.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+export let str = "Hello World";
+
+// main.js
+import { str } from './simple';
+console.log(str);
+```
+
+##### Babel
+
+To experiment with Babel I installed [babel-cli](https://babeljs.io/docs/usage/cli/) and configure it using a .babelrc config file. Then I run `npx babel-cli main.js simple.js -o output.js`.
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: Configure_d_

Comment (simonproffitt) [reply _posts/2017-11-21-javascript-modules.md]: ~run~ ran

File: _posts/2017-11-21-javascript-modules.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+});
+var str = exports.str = "Hello World";
+'use strict';
+
+var _simple = require('./simple');
+
+console.log(_simple.str);
+```
+
+Ah that's very interesting! ES6 export statements are added as properties to an object called `exports` while ES6 import use a function called `require`. However, none of them is defined in the output file so where are they coming from?! And this is where knowing a bit about the history of JavaScript modules is useful. Babel's job is to transpile to ES5 and since ES5 has no native modules, it converts ES Modules to one of the pre ES6 modules standard (e.g. CommonJS, AMD). By default, it uses the CommonJS modules standard and if you check the standard [here](http://wiki.commonjs.org/wiki/Modules/1.1) you will see that `exports` and `require` are indeed defined!
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: ~And~ This is where

File: _posts/2017-11-21-javascript-modules.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+  var str = exports.str = "Hello World";
+});
+define(['./simple'], function (_simple) {
+  'use strict';
+
+  console.log(_simple.str);
+});
+```
+
+So Babel transpiles ES6 modules to "something" well defined (AMD, CommonJS, etc...), this is great. However, where is the implementation of that "something" coming from? In the default configuration, `exports` and `require` must be defined somewhere, else whatever is executing my JavaScript (browser, node environment) won't be able to run it (except if there is some runtime magic I don't know about which would only half surprise me).
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: ~else~ otherwise

File: _posts/2017-11-21-javascript-modules.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+"use strict";
+Object.defineProperty(__webpack_exports__, "__esModule", { value: true });
+/* harmony export (binding) */ __webpack_require__.d(__webpack_exports__, "str", function() { return str; });
+let str = "Hello World";
+
+/***/ })
+/******/ ]);
+```
+
+And voilà! The missing piece of the puzzle is here. webpack is indeed the one taking care of defining or replacing `exports` and `require`. I will let you have fun reading webpack output (I debugged through the code in the browser, it was easier).
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: This was a little hard to understand as a non-native JavaScript speaker (probably not really your audience, I guess :)) Maybe worth breaking up the code or adding more comments to help explain what's going on?

File: _posts/2017-11-21-javascript-modules.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+
+#### Conclusion
+
+So here is what I understand about modules now (and remember to double check what I am writting): JavaScript modules exist natively only since [ES6](http://www.ecma-international.org/ecma-262/6.0/ECMA-262.pdf) (2015) so they are very recent. However, modules were added to the languages thanks to third party libraries and standards such as [AMD](https://github.com/amdjs/amdjs-api/blob/master/AMD.md) and [CommonJS Modules](http://wiki.commonjs.org/wiki/Modules/1.1). For web applications, since browsers do not widely support ES6 yet, ES6 modules must be transpiled to ES5 and some module implementation. In practice, when using [Babel](https://babeljs.io/) and [webpack](https://webpack.js.org/), Babel first transpiles ES6 modules to CommonJS modules (by default but can be changed) and webpack later converts it to its own module implementation.
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: Writing*

Comment (simonproffitt) [reply _posts/2017-11-21-javascript-modules.md]: JavaScript modules have _only_ existed natively ~only~ since

Comment (simonproffitt) [reply _posts/2017-11-21-javascript-modules.md]: By default but _this_ can be changed

**Review by simonproffitt — CHANGES_REQUESTED** [review]
This was really interesting, I didn't know anything about this topic at all! Have suggested a few bits of grammar and perhaps expanding on the webpack bit

File: _posts/2017-11-21-javascript-modules.md (hunk 1 above)
Comment (danabananarama) [inline _posts/2017-11-21-javascript-modules.md]: Maybe consider adding a brief note summarising what webpack and babel are? You kind of have it but it's not super clear to me, although i'm also quite new to JS stuff so might just be me (:

File: _posts/2017-11-21-javascript-modules.md (hunk 6 above)
Comment (danabananarama) [inline _posts/2017-11-21-javascript-modules.md]: 'none of them is' -> 'neither of them is'

**Review by danabananarama — COMMENTED** [review]
cool post, I should definitely refactor my massive scripts to use modules ^^;

File: _posts/2017-11-21-javascript-modules.md (hunk 6 above)
Comment (simonproffitt) [reply _posts/2017-11-21-javascript-modules.md]: Neither of them _are_

Comment (danabananarama) [reply _posts/2017-11-21-javascript-modules.md]: touché

File: _posts/2017-11-21-javascript-modules.md (hunk 2 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: oh I hadn't even thought about that haha

File: _posts/2017-11-21-javascript-modules.md (hunk 3 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: oh okay.

File: _posts/2017-11-21-javascript-modules.md (hunk 4 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: (I like that smiley so much: 🤔)

File: _posts/2017-11-21-javascript-modules.md (hunk 5 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: argh so sorry, you shouldn't have to correct that type of typo...

File: _posts/2017-11-21-javascript-modules.md (hunk 1 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: ok, I added very brief descriptions.

File: _posts/2017-11-21-javascript-modules.md (hunk 6 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: haha

File: _posts/2017-11-21-javascript-modules.md [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+  value: true
+});
+var str = exports.str = "Hello World";
+
+/***/ })
+/******/ ]);
+{% endhighlight %}
+
+The code is not complicated but not straightforward either if you don't debug through it. In a nutshell, webpack defines a function for each module (line 69 and 80) and stores them in an array. It then loads the first module, i.e. calls `__webpack_require__(0)` (line 64) which executes the first function in the array (line 21). Using this mechanism, webpack loads each module in the dependency graph (for instance module `1` (simple.js) is loaded by module `0` at line 74).
+
```
Comment (caroleolivier) [inline _posts/2017-11-21-javascript-modules.md]: I really struggle to explain the code :( let me know what you think, I am hoping it's a bit clearer.

**Comment by caroleolivier** [comment]
I noticed I made a mistake with the commands I was running. It doesn't change the command of my post but the code block were not entirely correct. I fixed all that.

File: _posts/2017-11-21-javascript-modules.md (hunk 10 above)
Comment (simonproffitt) [reply _posts/2017-11-21-javascript-modules.md]: It's clearer to me, the text has help me know where to look although in the `raw` view it doesn't render properly and I can't tell if that's an error in the Markdown or the viewer. This is a really interesting insight into whats's going on! 

I think it might be more correct to say 'step through it with a debugger' rather than 'debug through' but it's clear what you mean

File: _posts/2017-11-21-javascript-modules.md [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+In the JavaScript language, modules have only been available natively since 2015 (wow!) with the ECMAScript 6 [standard](http://www.ecma-international.org/ecma-262/6.0/ECMA-262.pdf). Not being able to break down code into modules is madness (in my opinion) and the community has come up with other ways to add modules to the language. (I am not really sure why modules were not added before, I don't know what the process is to create a new ECMAScript standard).
+<br/>
+The most famous module standards (pre ES6) are [AMD](https://github.com/amdjs/amdjs-api/blob/master/AMD.md)
+(a browser targeted standard whose most famous implementation is [requireJS](http://requirejs.org/)) and [CommonJS Modules](http://wiki.commonjs.org/wiki/Modules/1.1) (a standard defined by the group [CommonJS](https://en.wikipedia.org/wiki/CommonJS)).
+
+So depending on the context, people may refer to ES modules (ESM), CommonJS (CJS) modules or AMD (I don't think there is even consistency in acronym 🤔).
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: I still 🤔 this needs to be "acronyms"rather than "acronym" :)

File: _posts/2017-11-21-javascript-modules.md [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+
+#### Conclusion
+
+So here is what I understand about modules now (and remember to double check what I am writing): JavaScript modules have only existed since [ES6](http://www.ecma-international.org/ecma-262/6.0/ECMA-262.pdf) (2015) so they are very recent. However, modules were added to the languages thanks to third party libraries and standards such as [AMD](https://github.com/amdjs/amdjs-api/blob/master/AMD.md) and [CommonJS Modules](http://wiki.commonjs.org/wiki/Modules/1.1). For web applications, since browsers do not widely support ES6 yet, ES6 modules must be transpiled to ES5 and some module implementation. In practice, when using [Babel](https://babeljs.io/) and [webpack](https://webpack.js.org/), Babel first transpiles ES6 modules to CommonJS modules (by default but this can be changed) and webpack later converts it to its own module implementation.
```
Comment (simonproffitt) [inline _posts/2017-11-21-javascript-modules.md]: I wonder if this should be 'what I have written', using the past tense rather than what I am writing

**Review by simonproffitt — CHANGES_REQUESTED** [review]
A couple of small changes left but I really enjoyed the article - I love how it takes a simple example and traces through what's done to it by Babel and Webpack - I bet it's something loads of devs don't think about but I think I now have a good understanding having read the article

File: _posts/2017-11-21-javascript-modules.md (hunk 11 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: oops (haha)

File: _posts/2017-11-21-javascript-modules.md (hunk 12 above)
Comment (caroleolivier) [reply _posts/2017-11-21-javascript-modules.md]: agreed.

**Review by simonproffitt — APPROVED** [review]
Looks good to me, unless @kudakitsune spots anything I've missed
